### ISSUE TITLE
Derivation macros cause inexhaustive match warnings

### ISSUE THREAD

**Issue by mrdziuban** [Tinkoff/phobos#204]
Using scala 2.13 and the recently added `-Xlint:strict-unsealed-patmat` compiler option, the code generated by the derivation macros causes inexhaustive match warnings, seemingly related to a `match` statement using the field names of the given `case class`. Here's a scastie that reproduces the issue: https://scastie.scala-lang.org/mrdziuban/Ndn1ejJVQJeZ1OBXG9L6Tw/7

I'm able to work around the issue by adding a `-Wconf` compiler option to target and silence these specific warnings: 

```scala
scalacOptions += "-Wconf:msg=would fail on the following input.*x.*String forSome x not in.*macro:s"
```

but I was wondering if there was any way to add a fallback case to the `match` statement to make it exhaustive? Thanks in advance!

**Comment by valentiay** [comment]
Thanks for the bug report!

That's an interesting one. The pattern matching on case class constructor names has the [default clause](https://github.com/Tinkoff/phobos/blob/master/modules/core/src/main/scala-2/ru/tinkoff/phobos/derivation/DecoderDerivation.scala#L83), so this `match` should be exhaustive. I've faced some false positive inexhaustive match warnings in macros, while I was developing phobos for scala 3. Probably, this is a similar issue. I'll check if decoders are generated correctly, and I'll either fix the pattern matching or find a workaround to remove the warning.

Hope to resolve the issue this week

**Comment by mrdziuban** [comment]
Thanks for taking a look @valentiay! After adding the `-Ymacro-debug-lite` flag, I do see a default case generated by the code you linked to, but I think the one without a default case might be [this one](https://github.com/Tinkoff/phobos/blob/master/modules/core/src/main/scala-2/ru/tinkoff/phobos/derivation/DecoderDerivation.scala#L338-L341).

The full `-Ymacro-debug-lite` output is below, but just the relevant part is

```scala
case _root_.ru.tinkoff.phobos.derivation.DecoderDerivation.DecoderState.DecodingElement((name @ _)) => name match {
  case `s$macro$6` => {
    s$macro$3 = s$macro$3.decodeAsElement(cursor, s$macro$6, ElementCodecConfig.default.elementsDefaultNamespace);
    if (s$macro$3.isCompleted)
      s$macro$3.result({
        final <synthetic> <artifact> val rassoc$3 = s$macro$6;
        cursor.history.$colon$colon(rassoc$3)
      }) match {
        case _root_.scala.Right(_) => go(_root_.ru.tinkoff.phobos.derivation.DecoderDerivation.DecoderState.DecodingSelf)
        case _root_.scala.Left((error @ _)) => new _root_.ru.tinkoff.phobos.decoding.ElementDecoder.FailedDecoder[Playground.Test](error)
      }
    else
      go(_root_.ru.tinkoff.phobos.derivation.DecoderDerivation.DecoderState.DecodingElement(s$macro$6))
  }
}
```

<details>
  <summary>expand full output</summary>

```scala
{
  lazy val paramTypeclass$macro$7: ru.tinkoff.phobos.decoding.ElementDecoder[String] = decoding.this.ElementDecoder.stringDecoder;
  class ElementDecoder$macro$1 extends _root_.ru.tinkoff.phobos.decoding.ElementDecoder[Playground.Test] {
    <paramaccessor> private[this] val state: _root_.ru.tinkoff.phobos.derivation.DecoderDerivation.DecoderState = _;
    <paramaccessor> private[this] val s$macro$4: _root_.ru.tinkoff.phobos.derivation.CallByNeed[ru.tinkoff.phobos.decoding.ElementDecoder[String]] = _;
    def <init>(state: _root_.ru.tinkoff.phobos.derivation.DecoderDerivation.DecoderState, s$macro$4: _root_.ru.tinkoff.phobos.derivation.CallByNeed[ru.tinkoff.phobos.decoding.ElementDecoder[String]]) = {
      super.<init>();
      ()
    };
    def decodeAsElement(cursor: _root_.ru.tinkoff.phobos.decoding.Cursor, localName: _root_.java.lang.String, namespaceUri: _root_.scala.Option[_root_.java.lang.String]): _root_.ru.tinkoff.phobos.decoding.ElementDecoder[Playground.Test] = {
      var s$macro$3: ru.tinkoff.phobos.decoding.ElementDecoder[String] = s$macro$4.value;
      val s$macro$6 = ElementCodecConfig.default.transformElementNames("s");
      @new _root_.scala.annotation.tailrec() def go(currentState: _root_.ru.tinkoff.phobos.derivation.DecoderDerivation.DecoderState): _root_.ru.tinkoff.phobos.decoding.ElementDecoder[Playground.Test] = if (cursor.getEventType.$eq$eq(_root_.com.fasterxml.aalto.AsyncXMLStreamReader.EVENT_INCOMPLETE))
        {
          cursor.next();
          new ElementDecoder$macro$1(currentState, _root_.ru.tinkoff.phobos.derivation.CallByNeed[ru.tinkoff.phobos.decoding.ElementDecoder[String]](s$macro$3))
        }
      else
        currentState match {
          case _root_.ru.tinkoff.phobos.derivation.DecoderDerivation.DecoderState.New => if (cursor.isStartElement)
            _root_.ru.tinkoff.phobos.decoding.ElementDecoder.errorIfWrongName[Playground.Test](cursor, localName, namespaceUri) match {
              case _root_.scala.Some((error @ _)) => error
              case _root_.scala.None => {
                cursor.next();
                go(_root_.ru.tinkoff.phobos.derivation.DecoderDerivation.DecoderState.DecodingSelf)
              }
            }
          else
            new _root_.ru.tinkoff.phobos.decoding.ElementDecoder.FailedDecoder[Playground.Test](cursor.error("Illegal state: not START_ELEMENT"))
          case _root_.ru.tinkoff.phobos.derivation.DecoderDerivation.DecoderState.DecodingSelf => {
            if (cursor.isStartElement)
              cursor.getLocalName match {
                case `s$macro$6` => {
                  s$macro$3 = s$macro$3.decodeAsElement(cursor, s$macro$6, ElementCodecConfig.default.elementsDefaultNamespace);
                  if (s$macro$3.isCompleted)
                    s$macro$3.result({
                      final <synthetic> <artifact> val rassoc$3 = s$macro$6;
                      cursor.history.$colon$colon(rassoc$3)
                    }) match {
                      case _root_.scala.Right(_) => go(_root_.ru.tinkoff.phobos.derivation.DecoderDerivation.DecoderState.DecodingSelf)
                      case _root_.scala.Left((error @ _)) => new _root_.ru.tinkoff.phobos.decoding.ElementDecoder.FailedDecoder[Playground.Test](error)
                    }
                  else
                    go(_root_.ru.tinkoff.phobos.derivation.DecoderDerivation.DecoderState.DecodingElement(s$macro$6))
                }
                case _ => {
                  val state = _root_.ru.tinkoff.phobos.derivation.DecoderDerivation.DecoderState.IgnoringElement(cursor.getLocalName, Option(cursor.getNamespaceURI).filter(((x$1) => x$1.nonEmpty)), 0);
                  cursor.next();
                  go(state)
                }
              }
            else
              if (cursor.isEndElement)
                cursor.getLocalName match {
                  case (name @ _) if name.$eq$eq(localName) => s$macro$3.result({
  final <synthetic> <artifact> val rassoc$2 = s$macro$6;
  {
  final <synthetic> <artifact> val rassoc$1 = localName;
  cursor.history.$colon$colon(rassoc$1)
}.$colon$colon(rassoc$2)
}).map(((s$macro$5) => new Playground.Test(s$macro$5))) match {
                    case _root_.scala.Right((result @ _)) => {
                      cursor.next();
                      new _root_.ru.tinkoff.phobos.decoding.ElementDecoder.ConstDecoder[Playground.Test](result)
                    }
                    case _root_.scala.Left((error @ _)) => new _root_.ru.tinkoff.phobos.decoding.ElementDecoder.FailedDecoder[Playground.Test](error)
                  }
                  case _ => {
                    cursor.next();
                    go(_root_.ru.tinkoff.phobos.derivation.DecoderDerivation.DecoderState.DecodingSelf)
                  }
                }
              else
                {
                  cursor.next();
                  go(_root_.ru.tinkoff.phobos.derivation.DecoderDerivation.DecoderState.DecodingSelf)
                }
          }
          case _root_.ru.tinkoff.phobos.derivation.DecoderDerivation.DecoderState.DecodingElement((name @ _)) => name match {
            case `s$macro$6` => {
              s$macro$3 = s$macro$3.decodeAsElement(cursor, s$macro$6, ElementCodecConfig.default.elementsDefaultNamespace);
              if (s$macro$3.isCompleted)
                s$macro$3.result({
                  final <synthetic> <artifact> val rassoc$3 = s$macro$6;
                  cursor.history.$colon$colon(rassoc$3)
                }) match {
                  case _root_.scala.Right(_) => go(_root_.ru.tinkoff.phobos.derivation.DecoderDerivation.DecoderState.DecodingSelf)
                  case _root_.scala.Left((error @ _)) => new _root_.ru.tinkoff.phobos.decoding.ElementDecoder.FailedDecoder[Playground.Test](error)
                }
              else
                go(_root_.ru.tinkoff.phobos.derivation.DecoderDerivation.DecoderState.DecodingElement(s$macro$6))
            }
          }
          case _root_.ru.tinkoff.phobos.derivation.DecoderDerivation.DecoderState.IgnoringElement((name @ _), (namespace @ _), (depth @ _)) => if (cursor.isEndElement.$amp$amp(cursor.getLocalName.$eq$eq(name)).$amp$amp(cursor.getNamespaceURI.$eq$eq(namespace.getOrElse(""))))
            {
              cursor.next();
              if (depth.$eq$eq(0))
                go(_root_.ru.tinkoff.phobos.derivation.DecoderDerivation.DecoderState.DecodingSelf)
              else
                go(_root_.ru.tinkoff.phobos.derivation.DecoderDerivation.DecoderState.IgnoringElement(name, namespace, depth.$minus(1)))
            }
          else
            if (cursor.isStartElement.$amp$amp(cursor.getLocalName.$eq$eq(name)).$amp$amp(cursor.getNamespaceURI.$eq$eq(namespace.getOrElse(""))))
              {
                cursor.next();
                go(_root_.ru.tinkoff.phobos.derivation.DecoderDerivation.DecoderState.IgnoringElement(name, namespace, depth.$plus(1)))
              }
            else
              {
                cursor.next();
                go(currentState)
              }
        };
      go(state)
    };
    def result(history: _root_.scala.<byname>[_root_.scala.List[_root_.java.lang.String]]): _root_.scala.Either[_root_.ru.tinkoff.phobos.decoding.DecodingError, Playground.Test] = _root_.scala.Left(_root_.ru.tinkoff.phobos.decoding.ElementDecoder.decodingNotCompleteError(history));
    val isCompleted: _root_.scala.Boolean = false
  };
  new ElementDecoder$macro$1(_root_.ru.tinkoff.phobos.derivation.DecoderDerivation.DecoderState.New, _root_.ru.tinkoff.phobos.derivation.CallByNeed[ru.tinkoff.phobos.decoding.ElementDecoder[String]](paramTypeclass$macro$7))
}
```

</details>

**Comment by valentiay** [comment]
> I think the one without a default case might be [this one](https://github.com/Tinkoff/phobos/blob/master/modules/core/src/main/scala-2/ru/tinkoff/phobos/derivation/DecoderDerivation.scala#L338-L341).

Thanks! That's helpful. The problem is really in that pattern matching.

I've fixed in in #205 and published the fix in `0.15.1`

**Comment by mrdziuban** [comment]
Thank you!
